framework/include/neml2/userobjects/NEML2Assembly.h

```diff
@@ -0,0 +1,70 @@
+//* This file is part of the MOOSE framework
+//* https://mooseframework.inl.gov
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#ifdef NEML2_ENABLED
+
+// MOOSE includes
+#include "ElementUserObject.h"
+
+#include "neml2/tensors/Tensor.h"
+
+/// This user object caches assembly information from MOOSE.
+class NEML2Assembly : public ElementUserObject
+{
+public:
+  static InputParameters validParams();
+
+  NEML2Assembly(const InputParameters & parameters);
+
+  /// Number of active elements on this rank
+  int64_t numElem() const { return _nelem; }
+
+  /// Number of quadrature points per element
+  int64_t numQP() const { return _nqp; }
+
+  /**
+   * @brief Get the cached JxWxT for each element, each quadrature point
+   *
+   * The NEML2 tensor will have shape (nelem, nqp;), where nelem is the number of elements, and nqp
```

Comment:
```suggestion
   * The NEML2 tensor will have shape (nelem, nqp), where nelem is the number of elements, and nqp
```
or did you mean something by ;